test/e2e/gateway_api_test.go

```diff
@@ -0,0 +1,226 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorcontroller "github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller/gatewayclass"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/storage/names"
+
+	gwapi "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	// The expected OSSM subscription name.
+	expectedSubscriptionName = "servicemeshoperator"
+	// The expected OSSM catalog source name.
+	expectedCatalogSourceName = "redhat-operators"
+	// The expected catalog source namespace.
+	expectedCatalogSourceNamespace = "openshift-marketplace"
+	// The test gateway name used in multiple places.
+	testGatewayName = "test-gateway"
+)
+
+var crdNames = []string{
+	"gatewayclasses.gateway.networking.k8s.io",
+	"gateways.gateway.networking.k8s.io",
+	"httproutes.gateway.networking.k8s.io",
+	"referencegrants.gateway.networking.k8s.io",
+}
+
+// Global variables for testing.
+// The default route name to be constructed.
+var default_routename = ""
+
+// If the Gateway API feature gate is enabled, run a series of tests in order
+// to validate if Gateway API resources are available, objects can be created
+// successfully and also work properly, and that the Istio installation was
+// successful.
+// NOTE: do not change the name of the test.  If new tests are added while the
+// feature gate is still in effect, preface the test names with "TestGatewayAPI"
+// so that they run via the openshift/release test configuration.
+func TestGatewayAPI(t *testing.T) {
+	// Skip if feature is not enabled
+	if gatewayAPIEnabled, err := isFeatureGateEnabled(configv1.FeatureGateGatewayAPI); err != nil {
+		t.Fatalf("error checking feature gate enabled status: %v", err)
+	} else if !gatewayAPIEnabled {
+		t.Skip("Gateway API not enabled, skipping TestGatewayAPI")
+	}
+
+	// Defer the cleanup of the test gateway and special rolebinding.
+	t.Cleanup(func() {
+		testGateway := gwapi.Gateway{ObjectMeta: metav1.ObjectMeta{Name: testGatewayName, Namespace: operatorcontroller.DefaultOperandNamespace}}
+		if err := kclient.Delete(context.TODO(), &testGateway); err != nil {
+			if errors.IsNotFound(err) {
+				return
+			}
+			t.Errorf("failed to delete gateway %q: %v", testGateway.Name, err)
+		}
+		crb := &rbacv1.ClusterRoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "cluster-admin-e2e",
+			},
+		}
+		if err := kclient.Delete(context.TODO(), crb); err != nil {
+			if errors.IsNotFound(err) {
+				return
+			}
+			t.Errorf("failed to delete cluster role binding %q: %v", crb.Name, err)
+		}
```

Comment:
I notice we don't uninstall OSSM as a part of the cleanup. I think the philosophy is generally to cleanup any objects from tests to return the to the original state of the cluster.

One real-life problem being, while I'm testing this E2E locally, running the same test repeatedly changes the test. The 2nd time around, OSSM is already installed, so the installation won't get tested again. I imagine a scenario where we are debugging an OSSM installation flake and need to run it multiple times in a row. However, the test is perturbed after the first, so running multiple times is not testing the same environment.

But I understand this would be very tedious to code, I personally still can't figure out how to uninstall OSSM completely without breaking things. For now, what do you think about a simple future TODO?
```suggestion
                            // TODO: Uninstall OSSM after test is completed.
		}
```